src/realm/group_writer.cpp

```diff
@@ -992,9 +1092,50 @@ GroupWriter::FreeListElement GroupWriter::reserve_free_space(size_t size)
 {
     auto chunk = search_free_space_in_part_of_freelist(size);
     while (chunk == m_size_map.end()) {
-        // No free space, so we have to extend the file.
-        auto new_chunk = extend_free_space(size);
-        chunk = search_free_space_in_free_list_element(new_chunk, size);
+        if (!m_under_evacuation.empty()) {
+            // We have been too aggressive in setting the evacuation limit
+            // release chunks until we can fulfill the request
+            // Reverse sort according to position in file - we release from the back
+            std::sort(m_under_evacuation.begin(), m_under_evacuation.end(), [](const auto& a, const auto& b) {
+                return a.ref > b.ref;
+            });
+            bool found_or_give_up = false;
+            while (!found_or_give_up) {
+                auto& elem = m_under_evacuation.back();
+                if (elem.size >= size) {
+                    m_size_map.emplace(size, elem.ref);
+                    if (elem.size == size) {
+                        m_under_evacuation.pop_back();
+                    }
+                    else {
+                        elem.ref += size;
+                        elem.size -= size;
+                    }
+                    found_or_give_up = true;
+                }
+                else {
+                    // Chunk too small - just recycle so that we can try the next
+                    m_size_map.emplace(elem.size, elem.ref);
+                    m_under_evacuation.pop_back();
+                    found_or_give_up = m_under_evacuation.empty();
+                }
+            }
+            // Set new limit
+            if (m_under_evacuation.empty()) {
+                m_evacuation_limit = 0;
+                // std::cout << "Give up" << std::endl;
```

Comment:
Do we want to keep this commented `std::cout` lines... it is seems something left there after debugging…

Reply:
I think in this case, it will also help understand the code, and it makes it easier to reenable if you have to debug the code again.